File: code/backend/app/models/visual.py
"""
体块模式后端的请求/响应模型 (Step 11)。

四个对外接口对应四组 schema:
  POST /buildings/{id}/visual-models/block -> BlockModelRequest
  GET  /sites/{id}/scene                   -> SceneResponse
  GET  /buildings/{id}/visual-model         -> VisualModelOut
  DELETE /buildings/{id}/visual-models/{mid} (无 body)

不挂 response_model 到路由上: 项目统一响应 {code, message, data},
路由层挂 response_model 会和这个 wrapper 冲突 (见 retrieval.py 注释),
schema 只在 service 层组装时当结构定义用, 不挂到 FastAPI 上做自动序列化。

颜色 metric 字段约定:
  metric in {eui, total_kwh, anomaly_count}
  value 是原始值 (EUI kWh/m2 / 总能耗 kWh / 异常数), level 是分档后的字符串
  level in {low, mid, high, critical}

model_kind 优先级 splat > block, scene 返回每栋楼当前用的 kind:
  block     - 用户提交过体块参数 (DB 有 BLOCK 记录)
  splat     - 用户上传照片重建过 (DB 有 PHOTO_SINGLE / PHOTO_MULTI 记录)
  estimated - DB 没记录, service 层用 sqm + floors_count 自动估算
"""
from pydantic import BaseModel, Field, field_validator


class BlockModelRequest(BaseModel):
    """POST /buildings/{id}/visual-models/block 的 body。

    客户提交体块尺寸, 后端落 visual_model 记录 model_mode='BLOCK',
    render_format='BOX'。position 是园区内米制相对坐标, 可空 (没设过
    时前端布局算法兜底)。length/width/height 都必须 > 0, floors_count >= 1。
    yaw_deg 是水平旋转角度 (0-360 度), 可空 (默认 0 不旋转)。
    """

    length_m: float = Field(..., gt=0, le=100000, description="体块长度(米)")
    width_m: float = Field(..., gt=0, le=100000, description="体块宽度(米)")
    height_m: float = Field(..., gt=0, le=100000, description="体块总高(米)")
    floors_count: int = Field(..., ge=1, le=200, description="楼层数")
    position_x: float | None = Field(None, description="园区内 X 坐标(米), 可空")
    position_y: float | None = Field(None, description="园区内 Y 坐标(米), 可空")
    yaw_deg: float | None = Field(
        None, ge=0, lt=360, description="水平旋转角度(度, 0-360), 可空默认 0"
    )


class YawUpdateRequest(BaseModel):
    """PATCH /buildings/{id}/visual-models/yaw 的 body。

    只更新当前 active visual_model 的 yaw_deg, 不动其他字段。滑块拖动时
    前端 debounce 300ms 后调一次, 比走 POST /block 全量 upsert 轻。
    """

    yaw_deg: float = Field(..., ge=0, lt=360, description="水平旋转角度(度, 0-360)")


class BuildingUseUpdateRequest(BaseModel):
    """PATCH /buildings/{id}/use 的 body。

    更新 core.building 的用途字段, 决定 3D 页程序化表皮 + 屋顶类型。
    primary_use / sub_use 传 None 表示清空 (回到默认外观)。
    """

    primary_use: str | None = Field(None, description="主要用途 (如 Education / Commercial)")
    sub_use: str | None = Field(None, description="细分用途 (如 Academic / Retail / Residential)")


class Dimensions(BaseModel):
    """体块尺寸。统一给 scene API 和 building 详情接口用。"""

    length_m: float | None = None
    width_m: float | None = None
    height_m: float | None = None
    floors_count: int | None = None


class Position(BaseModel):
    """园区内相对坐标(米) + 水平旋转角度。"""

    x: float | None = None
    y: float | None = None
    yaw_deg: float | None = Field(
        None, description="水平旋转角度(度, 0-360), 默认 0 不旋转"
    )


class ColorMetric(BaseModel):
    """scene API 按 metric 算 value 和 level, 给前端 BoxGeometry 着色用。

    level 四档: low (绿) / mid (黄) / high (橙) / critical (红)。
    阈值在 config.py 里, 可调。
    """

    metric: str = Field(..., description="eui / total_kwh / anomaly_count")
    value: float | None = Field(None, description="metric 对应的原始值")
    level: str = Field(..., description="low / mid / high / critical")

    @field_validator("level")
    @classmethod
    def _check_level(cls, v: str) -> str:
        if v not in ("low", "mid", "high", "critical"):
            raise ValueError(f"level 取值非法: {v}")
        return v


class AnomalyStatus(BaseModel):
    """building 异常摘要, scene API 给前端在体块上叠角标用。

    severity_max 取 LOW/MEDIUM/HIGH 中最高的一档, 没异常时 None。
    by_severity 三档分别计数, 前端 tooltip 展开。
    """

    has_anomaly: bool
    severity_max: str | None = Field(None, description="LOW / MEDIUM / HIGH, 无异常 None")
    count: int = Field(0, description="异常总数")
    by_severity: dict[str, int] = Field(
        default_factory=lambda: {"LOW": 0, "MEDIUM": 0, "HIGH": 0},
        description="按严重度分组计数",
    )


class EnergyCompositionBadge(BaseModel):
    """building 能源构成小卡片。

    复用 query_service.get_building_energy_composition 的 composition 字段,
    结构 [{type, kwh, pct}], 给前端做小型堆叠条用。
    """

    composition: list[dict] = Field(default_factory=list)


class BuildingScene(BaseModel):
    """单栋楼在 scene API 里的完整元数据。

    model_kind 是场景渲染分流的关键字段:
      block     - 用 BoxGeometry 按 dimensions 渲染
      splat     - 用 splat loader 加载 .ply/.splat 文件 (storage_path 指向)
      estimated - 跟 block 一样渲染, 但前端可以加个虚线框表示是估算值
    """

    building_id: str
    building_code: str | None = None
    display_name: str | None = None
    model_kind: str = Field(..., description="block / splat / estimated")
    dimensions: Dimensions
    position: Position
    color_metric: ColorMetric
    anomaly_status: AnomalyStatus
    energy_composition: EnergyCompositionBadge
    model_id: str | None = Field(None, description="estimated 时为 None")


class SceneResponse(BaseModel):
    """GET /sites/{id}/scene 的响应。一次返整个园区的 3D 元数据。"""

    site_id: str
    time_range: dict
    metric: str
    buildings: list[BuildingScene]


class VisualModelOut(BaseModel):
    """GET /buildings/{id}/visual-model 的响应。

    单条 visual_model 记录的对外结构。没设过 visual_model 时路由层返 404,
    不会到这里。
    """

    id: str
    building_id: str
    model_mode: str = Field(..., description="BLOCK / PHOTO_SINGLE / PHOTO_MULTI / MANUAL_GLTF")
    render_format: str = Field(..., description="BOX / GLB / GLTF / PLY / SPLAT / SPZ")
    storage_path: str | None = None
    preview_image_path: str | None = None
    dimensions: Dimensions
    position: Position
    is_active: bool
    created_at: str

File: code/frontend/src/api/visual.ts
// ============================================================================
// Visual API - 建筑可视化模型 + 园区场景 + 重建 job
// ----------------------------------------------------------------------------
// 对齐后端 visual_model_service.list_site_scene 的实际返回结构 (嵌套):
//   site_id / time_range / metric / buildings[]
//   每栋 building 含 building_id / building_code / display_name / model_kind /
//   dimensions / position / color_metric / anomaly_status / energy_composition / model_id
//
// 路径: GET /api/v1/sites/{site_id}/scene?metric=&start=&end=
// ============================================================================

import { api } from './client'

// ---------------------------------------------------------------------------
// 类型定义 (对齐 backend/app/services/visual_model_service.py)
// ---------------------------------------------------------------------------

export type ModelKind = 'block' | 'splat' | 'estimated'
export type ColorLevel = 'low' | 'mid' | 'high' | 'critical'
export type AnomalySeverity = 'LOW' | 'MEDIUM' | 'HIGH'
export type SceneMetric = 'eui' | 'total_kwh' | 'anomaly_count'

export interface Dimensions {
  length_m: number | null
  width_m: number | null
  height_m: number | null
  floors_count: number | null
}

export interface Position {
  x: number | null
  y: number | null
  // 水平旋转角度 (度, 0-360), 默认 0 不旋转。后端 building_visual_model.yaw_deg。
  // BuildingSplat 渲染时把 yaw_deg 转弧度, 在 ENU 平移和 tile 局部缩放之间
  // 插入 R_z 旋转矩阵, 让 splat 绕"上"方向 (ENU Z 轴) 旋转。
  yaw_deg?: number | null
}

export interface ColorMetric {
  metric: SceneMetric
  value: number | null
  level: ColorLevel
}

export interface AnomalyStatus {
  has_anomaly: boolean
  severity_max: AnomalySeverity | null
  count: number
  by_severity: {
    LOW: number
    MEDIUM: number
    HIGH: number
  }
}

export interface EnergyCompositionItem {
  type: string
  kwh: number
  pct: number
}

export interface EnergyComposition {
  composition: EnergyCompositionItem[]
}

export interface SceneBuilding {
  building_id: string
  building_code: string
  display_name: string
  // 楼栋用途类型 (来自 core.building.primary_use, 如 Education / Science / Library)。
  // 前端 BuildingBlock 按这个选程序化表皮 + 屋顶类型, NULL 时走默认表皮。
  primary_use: string | null
  // 细分用途 (core.building.sub_use, 如 Academic / Sports Facility / Science
  // Facility / Student Center / Library)。比 primary_use 更细, 是选表皮的真正依据。
  sub_use: string | null
  model_kind: ModelKind
  dimensions: Dimensions
  position: Position
  color_metric: ColorMetric
  anomaly_status: AnomalyStatus
  energy_composition: EnergyComposition
  model_id: string | null
  // 后端 visual_model.tiles_path IS NOT NULL 时为 true。前端 BuildingSplat
  // 用这个判断走 Cesium3DTileset (原生 3DGS) 还是降级到 .ply 静态点云。
  has_tiles: boolean
}

export interface SceneResponse {
  site_id: string
  site_name: string
  time_range: { start: string; end: string }
  metric: SceneMetric
  buildings: SceneBuilding[]
}

// ---------------------------------------------------------------------------
// visual_model (单楼最新 active 模型)
//   GET /buildings/{id}/visual-model 返这个结构 (嵌套 dimensions / position)
//   用于 splat 模式拿 storage_path 拼 .ply URL
// ---------------------------------------------------------------------------

export type ModelMode = 'BLOCK' | 'PHOTO_SINGLE' | 'PHOTO_MULTI' | 'MANUAL_GLTF'
export type RenderFormat = 'BOX' | 'PLY' | 'GLTF'

export interface VisualModel {
  id: string
  building_id: string
  model_mode: ModelMode
  render_format: RenderFormat
  storage_path: string
  preview_image_path: string
  // 3D Tiles tileset.json 的绝对路径 (后端 worker 转 .ply -> 3D Tiles 后写入)。
  // NULL 表示未转换 (老记录 / 转换失败), 前端走 .ply 降级。
  tiles_path: string | null
  dimensions: Dimensions
  position: Position
  is_active: boolean
  created_at: string
}

// ---------------------------------------------------------------------------
// Reconstruction job
// ---------------------------------------------------------------------------

export interface ReconstructionJob {
  id: string
  building_id: string
  status: 'PENDING' | 'RUNNING' | 'SUCCEEDED' | 'FAILED'
  retry_count: number
  error_message: string | null
  started_at: string | null
  finished_at: string | null
  created_at: string
  output_model_id?: string | null
  dimensions?: Record<string, number>
  position?: Record<string, number | null>
}

export interface SinglePhotoRequest {
  photo_upload_id: string
  length_m: number
  width_m: number
  height_m: number
  floors_count: number
  position_x?: number
  position_y?: number
}

export interface BlockModelRequest {
  length_m: number
  width_m: number
  height_m: number
  floors_count: number
  position_x?: number | null
  position_y?: number | null
  yaw_deg?: number | null
}

// ---------------------------------------------------------------------------
// 3D Tiles URL helper
//   后端 /buildings/{id}/visual-models/{model_id}/splat-tiles/{file_path:path}
//   返 tileset.json + tiles/{level}/{x}.glb, 全部走 JWT 鉴权。
//
//   Cesium3DTileset.fromUrl 接受 string URL 或 Resource。用 Resource 时可设
//   headers, Authorization 会自动传给所有子 tile 请求 (tiles/{level}/{x}.glb)。
//   不在 URL 里塞 ?token=xxx: token 会泄漏到 referrer / nginx access log。
//
//   has_tiles 为 false (转换失败 / 老记录) 时返 null, 调用方降级走 .ply。
// ---------------------------------------------------------------------------

/**
 * 构造 Cesium3DTileset 用的 tileset.json URL (含 JWT 鉴权 header)。
 * 调用方喂给 `Cesium3DTileset.fromUrl(resource)`。
 *
 * 路径模式: `/api/v1/buildings/{building_id}/visual-models/{model_id}/splat-tiles/tileset.json`
 *
 * 用法:
 *   const resource = buildSplatTilesResource(buildingId, modelId)
 *   const tileset = await Cesium3DTileset.fromUrl(resource)
 *   viewer.scene.primitives.add(tileset)
 */
export function buildSplatTilesResource(buildingId: string, modelId: string): string {
  // 返 string URL, 让调用方自己构造 Cesium.Resource 加 headers。
  // 这里不直接 import Cesium 是为了保持 api/ 层纯逻辑 (无 Cesium 依赖)。
  return `/api/v1/buildings/${buildingId}/visual-models/${modelId}/splat-tiles/tileset.json`
}

// ---------------------------------------------------------------------------
// API 调用
// ---------------------------------------------------------------------------

export const visualApi = {
  /** 园区 3D 场景元数据。一次返整个园区所有 building 的渲染所需数据。 */
  getScene(siteId: string, params: { metric?: SceneMetric; start?: string; end?: string }) {
    return api.get<SceneResponse>(`/sites/${siteId}/scene`, { params })
  },

  /** 查 building 最新 active visual_model。splat 模式拿 storage_path 拼 .ply URL。
   *  没设过返 200 + null (后端把"没设过"当合法初始态, 不返 404)。 */
  getBuildingVisualModel(buildingId: string) {
    return api.get<VisualModel | null>(`/buildings/${buildingId}/visual-model`)
  },

  /** 提交体块参数。非源数据写操作, demo 也可用。 */
  submitBlockModel(buildingId: string, payload: BlockModelRequest) {
    return api.post<{ model_id: string; building_id: string; is_active: boolean }>(
      `/buildings/${buildingId}/visual-models/block`,
      payload,
    )
  },

  /** 更新建筑水平旋转角度 (Step 13)。前端 BuildingDetailDrawer 滑块拖动
   *  debounce 300ms 后调一次, 只 UPDATE 不 INSERT。返更新行数。 */
  updateBuildingYaw(buildingId: string, yawDeg: number) {
    return api.patch<{ building_id: string; yaw_deg: number; updated_count: number }>(
      `/buildings/${buildingId}/visual-models/yaw`,
      { yaw_deg: yawDeg },
    )
  },

  /** 更新建筑用途 (primary_use / sub_use), 决定 3D 页程序化表皮 + 屋顶类型。 */
  updateBuildingUse(
    buildingId: string,
    payload: { primary_use: string | null; sub_use: string | null },
  ) {
    return api.patch<{ building_id: string; primary_use: string | null; sub_use: string | null }>(
      `/buildings/${buildingId}/use`,
      payload,
    )
  },

  // ---- 重建 ----
  submitSinglePhotoJob(buildingId: string, payload: SinglePhotoRequest) {
    return api.post<{ job_id: string; status: string }>(
      `/buildings/${buildingId}/reconstruction/single-photo`,
      payload,
    )
  },

  getJobStatus(jobId: string) {
    return api.get<ReconstructionJob>(`/reconstruction/jobs/${jobId}/status`)
  },

  listJobs(params?: { building_id?: string; status?: string; limit?: number; offset?: number }) {
    return api.get<ReconstructionJob[]>('/reconstruction/jobs', { params })
  },

  retryJob(jobId: string) {
    return api.post<ReconstructionJob>(`/reconstruction/jobs/${jobId}/retry`)
  },

  deleteJob(jobId: string) {
    return api.delete<{ deleted: boolean }>(`/reconstruction/jobs/${jobId}`)
  },
}
